### ISSUE TITLE
Bounding threadpool threads/preventing deadlocks

### ISSUE THREAD

**Issue by PerilousApricot** [accre/lstore-gop#33]
Talked a bit with @tacketar on Skype about this.

Currently, attempting an operation like deleting a whole directory tree does the following (roughly):

```c
q = new_opque();
for (;;obj = tree->next()) { // sorted so the directory is last
    if is_dir(obj)
        opque_waitall(q);
    opque_add(delete_one_object_op(obj);
}
opque_waitall(q);
```

A few layers of indirection later, the delete_one_object_op(obj) ends up needing to remove all the IBP allocations in the lun segment:

```c
q = new_opque()
for (;;alloc = seg->next_alloc()) { // loop over each allocation
    opque_add(delete_one_alloc_op(alloc);
}
opque_waitall(q);
```

There are a couple of problems with this:
### This can deadlock.
If the upper layer completely fills up the thread pool, the lower layer can't run any jobs, which causes the whole threadpool to be filled with threads each doing ```opque_waitall```. Things work now because we try to have 10,000s threads in the pool, but that's roughly the same as putting in sleeps to prevent a race condition. GOP tries to make lemonade by turning ```opque_waitall```() into:
```c
if (op_not_executing(op)) {
   run inline;
} else {
   run async;
   _opque_waitall_real(op);
}
``` 
but that doesn't guarantee forward progress.
### Deadlocks are hard to prevent because GOP has no idea of what the true dependencies are.
In this example, there's the one toplevel opque and N lower queues, but nothing tying them directly together. If GOP knew which ops everyone was waiting on, it could be deadlock free by either either submitting them more intelligently or detect/fix deadlocks by checking for cycles and canceling threadpool tasks to break them. Instead, it takes all tasks from all queues and submits them identically to the same threadpool.
### Unilaterally (and unpredictably) converting ```opque_waitall(q)``` from an asynchronous scatter/gather into a serialized set of operations is probably not the right idea.
If the individual elements in the queue can block/hang, the whole queue will stall. (though, reading the code, I can't actually reason out a situation when this would happen. The serialization happens if there's both an element in the queue and the queue hasn't started running yet ([https://github.com/accre/lstore-gop/blob/master/gop.c#L623](here)). But, adding an element to the queue implicitly begins execution of the queue).
### A shedload of tasks/threads
Even assuming deadlocking is solved, we require at least one (if not many) tasks/threads for almost anything you would want to do. Here is the 'call stack' descending from lio_rm to actually removing something from IBP:
  - [lio_rm](https://github.com/accre/lstore-lio/blob/master/lio_rm.c#L125) - produces a queue
  - [lio_remove_regex_object](https://github.com/accre/lstore-lio/blob/master/lio_core_os.c#L470) - produces a queue
  - [lio_remove_object](https://github.com/accre/lstore-lio/blob/master/lio_core_os.c#L364) - synchronous
  - [exnode_remove](https://github.com/accre/lstore-lio/blob/master/exnode.c#L117) - produces a queue
  - [ds_remove](https://github.com/accre/lstore-lio/blob/master/segment_lun.c#L1904) - produces a queue
  - [modify_count_command](https://github.com/accre/lstore-ibp/blob/master/ibp_op.c#L1861) - directly calls
  - [send_command](https://github.com/accre/lstore-ibp/blob/master/ibp_op.c#L208) - directly calls
  - [gop_write_block](https://github.com/accre/lstore-ibp/blob/master/ibp_op.c#L861) - Hits network. Blocks on I/O

That's a lot, considering gop_write_block is the only person doing "real" work. Without any numbers, it's hard to hand wave what "a lot" is, but it's an interesting to think if having so many threads blocking on ```opque_waitall(q)``` and friends is the best way to structure this. It would be nice to quantify and profile that effect, especially to scale to potential new levels. Looking at what others have done, apache's thread-per-connection model is considered difficult to scale, and they put in a LOT of effort to try.

### Numbers would be nice
```Numbers talk, bullshit walks.``` If this is something considered important to tackle, it should be as data-driven as possible, no sense in falling into the "oh that seems cool" trap.

At the very least, it would be good to see:
1. The numbers of each type of tasks in the global task queue. Are most of the threads higher in the stack and waiting on their children, or are most of the threads down low and blocking on network I/O
2. At what rate are tasks being created/executed?
3. What are the mother/daughter dependencies between ops? Would it be possible to get a snapshot of every operation within the process, who produced them and who they produced? Along with a list of who they are blocking on? Can we get away with assuming the dependency graph is acyclic, or do we need to think about breaking cycles for deadlocks?

### Ideas
This is getting long in the tooth, but I have two ideas that might be useful to try and decrease the task usage. I'll post in a comment.

**Comment by PerilousApricot** [comment]
Two things that could help reduce the thread pressure:

### Tail calls
Often something like this happens in an op:
* make a queue with multiple ops
* wait for the whole queue to complete
* return the result of the queue to caller

This keeps a thread pool slot occupied. Instead, allow a lisp-esque tailcall:
* Init queue
* submit queue
* tell GOP the return value of my queue is my return value
* exit

### Waitall asynchronously
Another common pattern is:
* make a queue
* wait on queue
* do some thinking
* make another queue
* wait on queue

This is again occupying a thread which could be freed up. Instead, allow this pattern to be expressed as:

opA
* make queue
* tell GOP to update opB with the status of my queue

opB
* do some thinking
* make another queue
* wait on queue

This lets what would be a task which would naturally occupy a thread for a long time yield control back to GOP. That could help deadlock prevention

**Comment by tacketar** [comment]
I'm all for getting better bounds on the threadpools but it's the top 
level loop that causes the explosion.  The lower levels don't blossom 
like those.  Using the example you gave the 4 lower levels (ds_remove, 
modify_count_command, send_command, gop_write_block) will work as long 
as there are 2 threads available for the send/recv threads to a depot.  
It doesn't matter how many parallel tasks are being started at the top 
level.  Anything IBP related has hooks for  cap the total number of 
connections to the depots.  You can set the max number of connections 
and the max number per depot.  Giving it more threads just means it can 
work better.

A few concrete examples:

-lio_fuse running on a node with 12 CMS jobs uses 63 threads.
-lio_inspect repairing 40 files at a time uses 920 threads but I've got 
IBP configured to to use up to 512 connections (up to 1024 threads for 
IBP).  This is a hold over from when the depot network connection was 
buggy so we needed lots of connections to make things work.  I'll drop 
it down to 64 for the next repair. My guess is it will have 0 impact.

Alan

All the dependencies are definitely acyclic but they are not static.

**Comment by PerilousApricot** [comment]
Argh, trim responses! It detonates the web interface!

-- 
It's dark in this basement.

**Comment by PerilousApricot** [comment]
I think we're violently in agreement here: we should bring down the number of threads we need to be performant, and we should try and prevent resource starvation/priority inversion deadlocks. That's pretty uncontroversial.

I'm proposing we check and see who is blowing up. Running lio_rm at accre using the provided ulimit shouldn't lock like it does. I think it's all the bits in the middle, you think something else, but it's all bullshit until we dump out a trace and check.

If it *is* what I think, eliding the intermediate call stacks helps. If it's the lower layers, moving from thread-per-depot to a select-over-all-sockets system helps.

**Comment by tacketar** [comment]
I would add comments inline but it doesn't seem that that's a good 
idea.  Instead I'll just start fresh.

Tail Calls:
      I would be surprised if no extra processing was done before 
returning.  If nothing is being executed it should be possible to put 
the new queue on the parent gop/queue.  Otherwise ideally you'd like to 
to relinquish control of a waiting thread for other code to run.  It 
would also not require any code changes.  Just changes in the gop_wait 
routines.  This is also what you hint at for waitall async.

Waitall async:
     Manually splitting the code to do a task greatly increases the 
complexity.  I have to do it already in several locations.  It's not 
fun.  Makes the code much harder to understand and harder to debug.

Again hitting on my initial Tail call comment ideally you'd try and come 
up with a way to swap out the paused threads task to do more work.  Is 
that possible?  My gut says yes.

Alan

**Comment by PerilousApricot** [comment]
When I poked around, there was a lot of instances where basically it did exactly a tailcall modulo some memory freeing. Similarly, there's a decent number of loops that do "for opque_waitany(q): submit_op(q)".

Implementing the functionality in GOP doesn't mean everyone *has* to use it, I wouldn't waste the time unless it looked like the low hanging fruit, and you're right that it's a huge pain in the ass to get right.

But! Dead locking shouldn't happen. To me, the cost/benefit of "code locks" and "separating functions sucks" works.

**Comment by PerilousApricot** [comment]
And yes, you could do something where instead of waiting on a mutex/cond, you jump to another task, but usermode threading is a catastrophe compared with just avoiding waiting

**Comment by tacketar** [comment]
On 1/15/2016 4:04 PM, Andrew Melo wrote:
>
> When I poked around, there was a lot of instances where basically it 
> did exactly a tailcall modulo some memory freeing. Similarly, there's 
> a decent number of loops that do "for opque_waitany(q): submit_op(q)".
>

So who does the "modulo freeing some memory" stuff?  The devil is in the 
details.

Alan

**Comment by tacketar** [comment]
Outside of lio_rm?  Do you other examples that lock using the default 
nproc settings?  That one is definitely not thread friendly.  OTOH It's 
easy to cap nparallel to something small like 30.  My guess is doing 
that would make it work 100% of the time.

Alan

**Comment by PerilousApricot** [comment]
I meant "modulo freeing memory" to distinguish from the lisp semantics (the wait happens before the free instead of freeing before the weight). The ones I looked at could shuffle the frees fine. If not, GOP already does callbacks.

For another example, I don't have one handy.

In either case, I think (and want to show) nproc isn't the right limit. If I want to do 100 operations in a single process, I'd hope id not need 3,000 threads to get it done.

Either way, getting a good test case is a good way to answer this. maybe dropping nproc to 1 makes it not die, but it'll almost certainly suck. We can test the phase space. I'd naively think the number of threads should grow sub-linearly with the number or outstanding requests

**Comment by tacketar** [comment]
Another datapoint.  My inspect completed so I tweaked things and reran 
to see if it hurt performance and it didn't.  Dropping the number of IBP 
connections to 64 and/or upping the number of parallel inspects from 
40->100 I cap out at ~285 threads and I still get 10Gb/s line speed.  
Which got me thinking I've never seen the need for needing more the 
default 1k nproc except for deleting sometimes.... and realized I 
normally only had problems on Brazil.

Looking at brazil it's configured to use up to 4096 connections to the 
depots (8192 threads).  That's definitely not good.  If you used the 
clause from the cluster my guess is you'd cap out at about 300. Similar 
to the lio_inspect checks.

Alan

**Comment by PerilousApricot** [comment]
I was on the gateway. In either case, choosing the wrong number of threads should, at worst, slow things down.

I dunno - I ran something straightforward, it detonated, I looked into it, and it detonated for a really pathological reason. I want to find out the exact cause and solve it.

**Comment by tacketar** [comment]
Paul said you were supposed to be in town next week.  Let's try and 
reproduce your problem in the office.  I can't mange to blow the thread 
limit.  I've tried lio_cping 500 files at a time and also using lio_rm 
-np 500 to blow away 1000 files and I can't come close to hitting the 1k 
limit.  Wtih np=500 I can hit 650 threads on lio_cp (most of the time 
it's ~200) but with lio_rm I cap out at 267 threads.

Alan

**Comment by PerilousApricot** [comment]
Wanted to be in town next week. The ticket was literally 1/3 of my monthly gross pay, so I didn't go.

There or not, we can tickle it by making a terrible directory tree.

**Comment by PerilousApricot** [comment]
Initial commit of the bits to track the stack can be found at [feature/call-stack](https://github.com/accre/lstore-gop/tree/feature/call-stack).

Roughly, there is a thread-local variable pointing to the currently executing frame. Calling ```cs_frame_generic_init``` from the caller returns a new frame beneath the current frame and the new recursion depth. The helpers ```cs_frame_generic_begin``` and ```cs_frame_generic_end``` are called within the callee and set up/restores the thread local variables. For instance, the callee within a thread pool will have different TLS than the caller, so some magic has to be done to have the callee know who called it.

The frames are reference counted. I handle all of my references, leaving the reference from TLS. APR was taught how to decrement the count, so the remaining reference disappears automatically if either the TLS variable is changed or the thread terminates.
